### PR TITLE
Comprehensive refactor of bl-exit

### PR DESCRIPTION
**Pull request by tknomanzr**
This is a basic pull request. Shortcuts should be working now, although I am unsure if I fixed it or someone else, atm. I have refactored the code into modules to make it easier to read and digest. I have factored out bl_exit_merlin.py.deprecated although since we originally forked from this class, I left it in for reference purposes. The subsequent commits remove the .pyc files and remove the original bl-exit 

Signed-off-by: William Bradley <webradley9929@gmail.com>

### CODE REVIEW

File: bin/bl_exit.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+			windowicon = self.window.render_icon(gtk.STOCK_QUIT, gtk.ICON_SIZE_DIALOG)
+			self.window.set_icon(windowicon)
+
+		def configure(self):
+			if self.config_file:
+				try:
+					self.cp.read(self.config_file)
+				except ConfigParser.ParsingError as e:
+					# Log a [DEBUG} message here
+					print_message("{}: {}".format(__me__, str(e)))
```
Comment (ghost) [inline bin/bl_exit.py]: `__me__` is not in scope.

File: bin/bl-exit [hunk 2]
```diff
@@ -141,643 +43,126 @@ if display:
...
+							action="store_true")
+		parser.add_argument("-p", "--poweroff", help="Power off",
+							action="store_true")
+		parser.parse_args(sys.argv[1:])
+	"""No check if more than one option was specified. Take the first option and
+	discard the other"""
+	result = parser.parse_args()
+	return result
 
 def get_config_file():
```
Comment (ghost) [inline bin/bl-exit]: Let's rewrite this function to use python-xdg, it provides a more complete implementation and stable interface for querying canonical config directories. Also, this function completely fails on my system and returns None (Arch Linux, latest Python 2.7, KDE, XDG_CONFIG_HOME is unset), making the program crash. It remains to see why.

**Review by ghost — CHANGES_REQUESTED** [review]
Nice work. Besides some issues I've found while skimming the code (see the comments), I have some additional remarks:

 * Enforce a canonical module structure, and use setuptools (setup.py) as the canonical installation method. /debian/rules will then use pip for creating the Debian package. For this, let's use the module name bl_exit and go from there.

  * This requires fixing some imports, and concentrating the static assets in a folder like /bl_exit/assets/. I have sketched the structure I have in mind on this branch, based on your input: https://github.com/2ion/bunsen-exit/tree/tknomanzr. For providing the actual command, we can then use setuptool's console_scripts entry point feature (like @ https://github.com/2ion/gien/blob/master/setup.py#L18).

 * Stuff like .depracted files need to get removed from the tree (because of git, we still have the history anyway) or left in tree like `merlin.py`. Something else makes not much sense.

File: bin/bl_exit.py (hunk 1 above)
Comment (tknomanzr) [reply bin/bl_exit.py]: This sounds great. I am currently linting the code through pep8. As soon as I am done with that, I will work on these changes.

Comment (tknomanzr) [reply bin/bl_exit.py]: I am pretty sure I was able to cover this one in the subsequent pr. Please refer to the code in exit_config.py in my subsequent pull request.

**Comment by tknomanzr** [comment]
With your approval, we should probably close this pr and move onto the next one. I think I messed things up somehow. I know how to push to the helium-dev branch on my forked repository and issue a pull request but I did not know how to push those changes onto this pull request if that makes sense. I just wanted the conversation here to not disappear because there may be changes you requested here that I have not implemented in the refactored code yet.

**Comment by tknomanzr** [comment]
Consider this a proof of concept at this point. I still need to style our custom buttons as discussed on the forums, plus implement most of the details listed in bl-exitrc. 

This particular version is going to look for its config files in ~/.config/bunsen-exit as opposed to ~/.config/b-exit. This allows me to test the program without colliding with the current version of this program. Also, I have been making the naming more consistent where I can. If you run the program without access to any config files whatsoever, you should get an exit dialog that uses buttons with labels styled in the currently set gtk theme. However, if it can see the config files, it will run the currently unthemed version of our custom buttons, which are black and white at the moment.  
So here is my to-do:
- [x] - Style the custom buttons that I created. The good news here is styles can be applied to any Widget.
This will allow me to ditch the part of exit_config where I am attempting to manually read and parse those style.rc files.
- [x] Appy any leftover Theme entries from bl-exitrc that I have not yet implemented code for. This is mostly testing if the theme entry can be found in the theme_entries dictionary and applying those items where appropriate.
- [x] Continue testing and finding spots where I need try....catch blocks.
- [x] Implement some small changes that will allow us to select the level of logging detail from the command line. Right now it's an on/off sort of thing but I can tune the verbosity pretty easily.
- [x] I still need to implement show/hide functionality on the button labels as John requested.
- [ ] Finish any changes you requested that I have not yet addressed, for instance making a setup.py script for pip install functionality.
- [x] I will turn down the messages on the event handling soon. It is just there so that I could see that I was indeed grabbing the correct events. We don't need a log entry for every mouse_in and mouse_out event that gets triggered.

The structure of bl-exitrc has changed in small ways. In particular, the button entries were reworked. The button_entries dictionary pretty much maps the entire dialog in terms fo which buttons to create.  Additionally, I have placed a couple of small changes to helium.rc. Tooltips may need a bit of additional work. Changing text color on labels has the side effect of also changing it on tooltips when using gtk.styleRc. 

Buttons now highlight correctly. That was one of those 4 a.m. inspirations. By setting the focus on the button on mouse_in, it removes the problem of double highlights. (NOTE: I would not ever do this for a data entry application but it should be fine for this application). I think the colored button class is inheriting behaviors and attributes from the gtk.Widget class, which is good because that gives us the ability to style the buttons using gtk.styleRc. If not, then it will and I will pack the event box into the widget, etc. 
 
NOTE: The entry point for the application is now bunsen-exit. You can run the command with the command ./bunsen-exit. It has been made executable.

**Comment by ghost** [comment]
Thank you very much for the update! I promise I'll try and make some room this week to help you on this.

**Comment by tknomanzr** [comment]
If I modify ~/.config/bunsen-exit/bl-exitrc and add the following to it:
```
window_background_normal=#284F63
button_background_normal=#73919C
button_background_prelight=#678B8B
text_color_normal=#00ffff
text_color_prelight=#000000
```

it becomes easy to map the background colors to the custom ColoredImageButtons for the various events that we care about, i.e mouse_in, mouse_out, focus_in, and focus_out. This method bypasses the gtkRcStyles settings altogether. Text color then has to be set via Pango markup. We could then also specify a font setting in our bl-exitrc file to set a custom font on the button Labels. I had to redo the tooltips for these custom buttons somewhat as well.

This push is pretty close. I just have a few more settings to go, depending on what else I end up adding. I am thinking if we go down this route, in order to keep things modular, we would need to split the bl-exitrc up some. What I am thinking is that file sets some meta-data, i.e. what buttons to show, theme name, theme author, which style to set. The rest specifies appearance. If we break out the appearance stuff into a file named helium.rc for example, I could then modify the class that pulls all those settings in to read that separate file. Basically, I want to replace the rc file with an ini styled file that reads in the same settings. This would allow me to keep things modular while at the same time bypassing the gtkRcStyles. 

I can get the style to take for standard items like a gtk.Button but for the class that I created, which basically packs some widgets into a gtk.EventBox, I am having trouble figuring out how to set a style on it. 
This push resolves the problem. However, I do not know how committed you are to keeping the styles as-is.

File: exit_gtk.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+			_, _, screen_width, _ = screen.get_monitor_geometry(curmon)
+		except:
+			exit_log.info('Error constructing UI. Not running under X')
+		finally:
+			del x, y, display, screen, curmon
+		exit_log.debug("Detected screen_width is " + str(screen_width))
+		try:
+			self.width_adjustment = float(self.theme_entries['window_width_adjustment'])
+		except:
+			exit_log.warn("window_width_adjustment not a float. Please check config. Defaulting to 0.5")
```
Comment (ghost) [inline exit_gtk.py]: Missing `self.width_adjustment = 0.5` in `except:` block, leading to segfault for me in the following line when the value is being accessed.

File: exit_gtk.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+		self.window.connect("delete_event", self.destroy)
+		self.window.connect("destroy_event", self.destroy)
+		self.window.connect("key-press-event", self._key_press_event)
+		self.window.set_resizable(False)
+		self.window.set_keep_above(True)
+		self.window.stick()
+		self.window.set_position(gtk.WIN_POS_CENTER)
+		windowicon = self.window.render_icon(gtk.STOCK_QUIT, gtk.ICON_SIZE_DIALOG)
+		self.window.set_icon(windowicon)
+		self.window.modify_bg(gtk.STATE_NORMAL, gtk.gdk.color_parse(self.theme_entries['window_background_normal']))
```
Comment (ghost) [inline exit_gtk.py]: Similar issue. On my Arch system (bunsen-exit never ran here), this leads to a gtk segfault because `self.theme_entries` does not (yet) have a key `window_backgrund_normal`. Trace:
```
__init__.py(1341):         if (found == 0) and raiseExceptions and not self.manager.emittedNoHandlerWarning:
exit_gtk.py(88):                self.window.set_name('Bunsen Exit')
exit_gtk.py(89):                self.window.set_decorated(False)
exit_gtk.py(90):                self.window.connect("delete_event", self.destroy)
exit_gtk.py(91):                self.window.connect("destroy_event", self.destroy)
exit_gtk.py(92):                self.window.connect("key-press-event", self._key_press_event)
exit_gtk.py(93):                self.window.set_resizable(False)
exit_gtk.py(94):                self.window.set_keep_above(True)
exit_gtk.py(95):                self.window.stick()
exit_gtk.py(96):                self.window.set_position(gtk.WIN_POS_CENTER)
exit_gtk.py(97):                windowicon = self.window.render_icon(gtk.STOCK_QUIT, gtk.ICON_SIZE_DIALOG)
exit_gtk.py(98):                self.window.set_icon(windowicon)
exit_gtk.py(99):                self.window.modify_bg(gtk.STATE_NORMAL, gtk.gdk.color_parse(self.theme_entries['window_background_normal']))
 --- modulename: trace, funcname: _unsettrace
trace.py(80):         sys.settrace(None)
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/usr/lib/python2.7/trace.py", line 819, in <module>
    main()
  File "/usr/lib/python2.7/trace.py", line 807, in main
    t.runctx(code, globs, globs)
  File "/usr/lib/python2.7/trace.py", line 513, in runctx
    exec cmd in globals, locals
  File "bunsen-exit", line 160, in <module>
    sys.exit(main())
  File "bunsen-exit", line 154, in main
    bl_exit_gtk = exit_gtk.ExitGtk(button_values, exit_bus, theme, theme_entries, style_path)
  File "exit_gtk.py", line 99, in __init__
    self.window.modify_bg(gtk.STATE_NORMAL, gtk.gdk.color_parse(self.theme_entries['window_background_normal']))
```

Comment (ghost) [reply exit_gtk.py]: FYI: How I obtained those traces:
```
python2.7 -m trace --trace ./bunsen-exit
```

File: exit_gtk.py (hunk 3 above)
Comment (tknomanzr) [reply exit_gtk.py]: If no configuration can be found for whatever reason, you should now get a dialog built using the default gtk theme for the system. If one can be found, then testing that each value is valid and if not, set to some sane default is still a work-in-progress. Also, the configuration files are still in a state of flux due to the fact that I have been mapping in functionality as I go along. However, at the moment, the program is looking for bl-exitrc in ~/.config/bunsen-exit. 

I noticed on the custom buttons, that the enter key was not being handled such that tab...tab..enter was activating the button as one would expect. The lastest commit address that as well. I will be working on this pretty heavily for the next couple of days so expect more incoming commits.

Comment (ghost) [reply exit_gtk.py]: Yup, now it works. Thanks.

Comment (johnraff) [reply exit_gtk.py]: This looks excellent - thank you!

> the configuration files are still in a state of flux due to the fact that I have been mapping in functionality as I go along. However, at the moment, the program is looking for bl-exitrc in ~/.config/bunsen-exit.

I'm wondering about backwards compatibility here. Is there any possibility of having old-style bl-exitrc files interpreted to some extent? (The same "progressive enhancement" philosophy you applied with falling back on a default style when there's no config file at all.)

Comment (tknomanzr) [reply exit_gtk.py]: I have been planning to do some refactoring on all of the code that tests for valid configuration entries and sets them to some sane default, quite possibly the stock BL look for Hydrogen, since that is where we started and the old #! look would be better covered by using a #! gtk theme and no config file at all.  At the moment, those tests are scattered all throughout the program. I want to group them all up and bring them back under the exit_config module.py. With these tests, if an option cannot be parsed, a sensible default will be set and if logging is enabled, an entry will be written with a [WARN] tag.

So my plan is to group the logging levels as such:
[INFO] General and fairly brief info, such as logging started, which configuration is being loaded, etc.
[WARN] This logging level would be for modders. A warning entry means that a configuration item could not be parsed and so the program is setting some default as I mentioned above. This would be our way of alerting those who wish to mod the program, "Hey I tried to parse this configuration item but could not so I set it to something meaningful."
[DEBUG] This level is basically full introspection of all the inputs and outputs of the program. It is very verbose but gives a good view of what is going on in the program.

In the case that the program attempts to parse an old bl-exitrc, the button entries have changed. The old style looks like this:
```
[cancel]
show = always
onerror = novisual

[logout]
show = always
onerror = novisual

[suspend]
show = always
onerror = novisual

[sleep]
show = always
onerror = novisual

[hibernate]
show = never
onerror = novisual

[hybridsleep]
show = never
onerror = novisual

[reboot]
show = always
onerror = novisual

[poweroff]
show = always
onerror = novisual
```
While the new one looks like this:
```
[button_values]
Cancel = show
Logout = show
Suspend = show
Reboot = show
PowerOff = show
Hibernate = show
HybridSleep = show
`````
The onerror stuff has been replaced by a different logging mechanism. Additionally, I don't pop up error dialogs anywhere. You remember the Windows donk with the blocking dialog ala Windows 95 - 98? Yeah, I hate those things.

What should happen is that the program will decide it can't use the button array as presented in the old bl-exitrc and build a default button list which shows every button except Hibernate and HybridSleep by default. I can place a log message that tells the user is cannot parse the button array, most likely because it is using an old configuration format and set the default array, while encouraging them to update their configuration file.

Similarly, if the theme specified cannot be found, it will build an exit dialog based on some set of defaults (I mentioned Hydrogen's look earlier.) This behavior will continue for each theme_entry in the specified theme, each time issuing a [WARN] in the logs if it overrides some theme_entry.

To summarize, I can ensure that the program will build some sort of default themed exit dialog if it finds an old style bl-exitrc. I can also issue [WARN] logs that encourage folks to use the new configuration setup. The new configuration setup, I hope, can become a tool that our modding community can use in their desktop builds. There is additional functionality that I have not mapped in yet but is possible, most notably, the ability to build button borders and the ability to set fonts and their associated sizes for the button labels. 

Finally, I will ensure that the new configuration files end up heavily commented so that it will hopefully be apparent to folks how to use the config files to build their own exit dialog themes.

I hope this will serve to alleviate your concerns somewhat.

Comment (johnraff) [reply exit_gtk.py]: > I hope this will serve to alleviate your concerns somewhat.

Yes, indeed. Thank you!

**Comment by tknomanzr** [comment]
I am considering bunsen-exit to be more or less complete now. There are some features that would be nice to have that we might want to work in later but I believe the program fulfills the requirements that you presented to me. Now I will start testing it and see how many ways I can break it.

**Comment by tknomanzr** [comment]
## Tests from helium.rc completed:
### name:
- Missing value: Passed
- Missing key: Passed
### author:
- Missing value: Passed
- Missing key: Passed
### window_width_adjustment:
- Missing value: Passed
- Missing key: Passed
- Value not a float: Passed
### dialog_height:
- Missing Value: Passed
- Missing key: Passed
- Value not an int: Passed
### inner_border:
- Missing value: Passed
- Missing key: Passed
- Value not an int: Passed
### sleep_delay:
- Missing value: Passed
- Missing key: Passed
- Value not a float: Passed
### overall_opacity:
- Missing value: Passed
- Missing key: Passed
- Value not an int: Passed
### button_height:
- Missing value: Passed
- missing key: Passed
- value not an int: Passed
### button_spacing:
- Missing value: Passed
- missing key: Passed
- value not an int: Passed
### icon_path:
- Missing value: Passed
- Missing key: Passed
- directory not found: Passed and causes button images to default to gtk.STOCK_ERROR_DIALOG, since path is not valid.
### button_image_cancel:
- Missing value: Passed
- Missing key: Passed
- file not found: Presents gtk.STOCK_ERROR_DIALOG in ui.
### button_image_poweroff:
- Missing value: Passed
- Missing key: passed
- file not found: Presents gtk.STOCK_ERROR_DIALOG in ui.
### button_image_reboot:
- Missing value: Passed
- Missing key: Passed
- file not found: Presents gtk.STOCK_ERROR_DIALOG in ui.
### button_image_suspend:
- Missing value: Passed
- Missing key: Passed
- file not found: Presents gtk.STOCK_ERROR_DIALOG in ui.
### button_image_logout:
- Missing value: Passed
- Missing key: Passed
- file not found: Presents gtk.STOCK_ERROR_DIALOG in ui.
### button_image_hybridsleep:
- Missing value: Passed
- Missing key: Passed 
- file not found: Presents gtk.STOCK_ERROR_DIALOG in ui.
### button_image_hibernate:
- Missing value: Passed
- Missing key: Passed
- File not found: Presents gtk.STOCK_ERROR_DIALOG in ui.
### label_height:
- Missing value: Passed
- Missing key: Passed
- Value not an int: Passed
### window_background_normal:
- Missing value: Passed
- Missing key: Passed
- Attempt to gtk.gdk.color_parse value: Passed
### button_background_normal:
- Missing value: Passed
- Missing key: Passed
- Attempt to gtk.gdk.color_parse value: Passed
### button_background_prelight:
- Missing value: Passed
- Missing key: Passed
- Attempt to gtk.gdk.color_parse value: Passed
### text_color_normal:
- Missing value: Passed
- Missing key: Passed
- Attempt to gtk.gdk.color_parse value: Passed
### text_color_prelight:
- Missing value: Passed
- Missing key: Passed
- Attempt to gtk.gdk.color_parse value: Passed
### tooltip_background:
- Missing value: Passed
- Missing key: Passed
- Attempt to gtk.gdk.color_parse value: Passed
### tooltip_foreground:
- Missing value: Passed
- Missing key: Passed
- Attempt to gtk.gdk.color_parse value: Passed
### font_family:
- Missing value: Passed
- Missing key: Passed
### font_style:
- Missing value: Passed
- Missing key: Passed
### font_size:
- Missing value: Passed
- Missing key: Passed
### tooltip_font_family:
- Missing value: Passed
- Missing key: Passed
### tooltip_font_style:
- Missing value: Passed
- Missing key: Passed
### tooltip_font_size:
- Missing value: Passed
- Missing key: Passed

**Comment by johnraff** [comment]
@tknomanzr You might find @2ion 's apt-sourcemgr package of interest in terms of python packaging. In particular, building the man page: https://github.com/BunsenLabs/apt-sourcemgr/commit/f69f349c5b2a846131fc65bb28b62b7ccc52635e

**Comment by tknomanzr** [comment]
the command

> sudo python setup.py install

now works correctly. It installs the binary bl-exit into /usr/local/bin.
Run it with the command

> /usr/local/bin/bl-exit
